Add `accelerate` support for ViLT

younesbelkada opened the pull request:
# Motivation

Add `bnb` support for ViLT model as it has been asked by a user in https://github.com/TimDettmers/bitsandbytes/issues/14 
. This involved adding `accelerate` support for this model.

# What does this PR do?

Adds `_no_split_modules` attribute at `ViltModel` class to support  loading the model with `device_map=auto`. This also implied adding a `.to` operation inside `ViltLayer`. 
I also redefined `accelerate` tests since for this model the hidden states are not deterministic. However, it is possible to check the correctness of the operation by checking some output attributes such as `logits` or `pooler_output`. 

# Questions

The test `ViltModelIntegrationTest::test_inference_natural_language_visual_reasoning` seem to never pass on my machine (aka even without `_no_split_modules`), is it related to something I am missing? Also it seems that those tests were failing too on the nightly run: https://github.com/huggingface/transformers/runs/7882898294?check_suite_focus=true 

cc @NielsRogge @ArthurZucker @ydshieh

younesbelkada committed:
add `accelerate` support for ViLT

- redefine  `accelerate` tests by picking the correct model output
- redefined tests tolerance due to stochasticity
- slow tests are passing except `iltModelIntegrationTest::test_inference_natural_language_visual_reasoning`
- But the test above seem to never pass anyway

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into vilt-accelerate-suuport

ydshieh commented:
> I also redefined accelerate tests since for this model the hidden states are not deterministic.

Could you elaborate this part in more depth, please?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18683). All of your documentation changes will be reflected on that endpoint.

younesbelkada committed:
fix typo

younesbelkada commented:
No worries! 
After deeply looking at the code it appears that the `pooler` and the other heads that are used by the model takes as an input the [first element](https://github.com/huggingface/transformers/blob/358fc18613a737f3fbcebc5b2abed43386ff9cbc/src/transformers/models/vilt/modeling_vilt.py#L885) of the variable `hidden_states` (with the latest having the shape `batch_size`x `nb_tokens` x `hidden_dim`). With the hidden states containing an embedding for each image patch ([stochastic](https://github.com/huggingface/transformers/blob/358fc18613a737f3fbcebc5b2abed43386ff9cbc/src/transformers/models/vilt/modeling_vilt.py#L174)) [concatenated together with the text embeddings (non stochastic)](https://github.com/huggingface/transformers/blob/358fc18613a737f3fbcebc5b2abed43386ff9cbc/src/transformers/models/vilt/modeling_vilt.py#L236). Therefore since the heads uses the first text embedding (or CLS token) to perform its downstream task, there is no stochasticity involved there. The stochasticity is only involved on the image patch embedding side.
This can be confirmed as well from the model achitecture (taken from [https://arxiv.org/pdf/2102.03334.pdf](arxiv)):
![Screenshot 2022-08-18 at 17 27 05](https://user-images.githubusercontent.com/49240599/185434093-1b6ca99e-8fb2-4353-9861-a382f675b9ac.png)
Although one can always take the first hidden states to perform the test, I do think the proposed fix is slightly better since it checks the output of modules that would not be checked if we consider only the first hidden states.

younesbelkada commented:
After looking into it with @ArthurZucker fixing a manual seed on each accelerate test before each forward pass seems to fix the non passing tests. 
We can either keep the changes as they are, or re-define the tests with a seed that is set before each forward pass

ydshieh commented:
I am open to the fix of using seed. It makes the change smaller.
But I am confused by the fact that the daily scheduled CI never report this issue (I should check on slack channels too).

https://github.com/huggingface/transformers/runs/7891383348?check_suite_focus=true

The nightly CI you mentioned above uses nightly torch version (i.e. daily built torch) for which more test failures are expected.

ydshieh commented:
I think it's better for us to figure out why the relevant tests don't fail on scheduled CI for a long period, but here it fails. This seems contradictory.

ydshieh commented:
Could you mention on which hardware you tested, @younesbelkada ?

younesbelkada committed:
few modifications

- replace everything by manually setting the seed

younesbelkada commented:
For the `accelerate` tests it is normal that they were passing since the `_no_split_modules` class was never define in the model, therefore these tests were never run. 
Regarding the second test maybe it's a hardware issue yes! Let me print you the details:
```
Python version: 3.10.4 (main, Mar 31 2022, 08:41:55) [GCC 7.5.0]
transformers version: 4.22.0.dev0
Torch version: 1.12.1
Cuda available: True
Cuda version: 11.3
CuDNN version: 8302
Number of GPUs available: 2
NCCL version: (2, 10, 3)
DeepSpeed version: None
TensorFlow version: 2.9.1
TF GPUs available: True
Number of TF GPUs available: 2
```

younesbelkada commented:
The non passing test pass on my VM with `atol=4e-2` by the way

younesbelkada commented on tests/models/vilt/test_modeling_vilt.py:
```diff
@@ -530,7 +668,8 @@ def test_for_token_classification(self):
 
 # We will verify our results on an image of cute cats
 def prepare_img():
-    image = Image.open("./tests/fixtures/tests_samples/COCO/000000039769.png")
```
@ydshieh maybe it's because of this line? Is `./tests/fixtures/tests_samples/COCO/000000039769.png` stored in the Docker image?

ydshieh replied:
It's in the repo, no? The CI clones the repo., so it should be there.

Thanks for the `_no_split_modules` not set part, now clear to me! So good for me to use seed for those tests.

younesbelkada replied:
Great my bad then, reverted that change. However this seems to not fix the initial issue

younesbelkada replied:
It's probably hardware / library version related, I saw somewhere that different PIL version can yield to different results

younesbelkada replied:
Check here: https://github.com/huggingface/transformers/blob/bbbb453e5869a5fec4925b02f1265c9e6bfc3ebb/tests/models/vilt/test_modeling_vilt.py#L637

ydshieh replied:
By `the initial issue`,  you mean `ViltModelIntegrationTest::test_inference_natural_language_visual_reasoning`?
As it works on CI, I won't worry too much. I always try to Nvidia T4.

ydshieh replied:
I am not sure why you changed th line regarding image though.

younesbelkada replied:
Yes I meant `ViltModelIntegrationTest::test_inference_natural_language_visual_reasoning` 
Just because I was running my tests not in the parent folder but from somewhere else, now that I am running the tests directly from the parent folder I prefer to keep the old changes

ydshieh replied:
Pillow versions indeed counts (in general) - we had some test failures before due to this.

younesbelkada replied:
Great then I guess we can assume it is library version related! 💪 
Let me know if there is anything else which is unclear!

ydshieh replied:
Could we just set the seed, then call `super().test_cpu_offload()`?

younesbelkada replied:
The only thing I am worried is that you need to set a seed before each inference step, so 2 times. Once before [the first forward pass](https://github.com/younesbelkada/transformers/blob/6a33f8a6388b24a9463a47a1822b5bff321bb3c7/tests/models/vilt/test_modeling_vilt.py#L543) and the second time before the [other forward passes ](https://github.com/younesbelkada/transformers/blob/6a33f8a6388b24a9463a47a1822b5bff321bb3c7/tests/models/vilt/test_modeling_vilt.py#L560) that is why I think that we cannot call `super().test_cpu_offload()` but have to define our own function

younesbelkada replied:
Maybe there is a way to fix a seed without having to re-set it at each inference step

younesbelkada replied:
Let me check

younesbelkada replied:
I propose to add a context manager at 3cda114517680c8c215f82d90ff8a0a7b8844523
with that there is no need to redefine the whole function. We just need a decorator `set_reproducible` to be set on these functions.

younesbelkada committed:
revert local image

younesbelkada committed:
add context manager instead

younesbelkada committed:
fix comment

NielsRogge commented on src/transformers/testing_utils.py:
```diff
@@ -1615,3 +1616,38 @@ def new_request(self, method, **kwargs):
+
+
+# adapted from https://stackoverflow.com/questions/32163436/python-decorator-for-printing-every-line-executed-by-a-function
+class set_reproducible(ContextDecorator):
```
This is interesting, does that mean we can use this decorator every time we want to test PyTorch models that involve stochasticity? Cause ViTMAE is another example (as it creates a random boolean mask inside).

I'm currently using `torch.manual_seed` as seen [here](https://github.com/huggingface/transformers/blob/0f257a87749e0a72bda260c6f319a45dae1e7c4d/tests/models/vit_mae/test_modeling_vit_mae.py#L235). Should we use this decorator instead?

ydshieh replied:
I think this change could be well applied to `ViTMAE` - they have exactly the same stochasticity situation.

ArthurZucker replied:
It does work for the mentioned test 🥳

younesbelkada committed:
Merge branch 'main' into vilt-accelerate-suuport

younesbelkada commented:
Thanks a lot for your comment @ydshieh & @NielsRogge 

@NielsRogge : I can confirm this decorator works fine for ViTMAE too, I replaced the function you pointed me by:
```
@set_reproducible(2)
    def test_save_load(self):
        super().test_save_load()
```
And the test was passing so I guess we can safely replace stochastic tests with this kind of trick

I would love to have a quick review from @sgugger or @stas00 if possible 🙏 As I think that this decorator can be useful for future models

Thanks 💪

ArthurZucker commented:
Just a small comment, in terms of performances I think the decorator can be a little bit improve to only run on the model's forward and not on every single forward pass (if I understand correctly here all the functions named `forward` are affected, instead of just the model's main forward pass)

sgugger commented on src/transformers/testing_utils.py:
```diff
@@ -1615,3 +1616,38 @@ def new_request(self, method, **kwargs):
+
+    def set_seed(self, frame, event, arg):
+        # Set the seed
+        torch.manual_seed(self.seed)
```
Can we maybe use the `set_seed` function here from our utils (so it sets all seeds and not just PyTorch, while also working when PyTorch is not installed)?

younesbelkada replied:
It seems that yes we can! added the proposed change in cb7a4029794427dfeb78f1c4cf50a277f709289d

sgugger reviewed (approved):
Very nice new util! Thanks a lot for adding this and making the tests pass for VilT!

younesbelkada committed:
small fix

- use `transformers` set_seed instead of `torch.manual_seed`

younesbelkada commented:
I think that @ArthurZucker wanted to enhance the util to make it more memory efficient, will let him decide whenever he feels it ready to merge 💪

ydshieh reviewed (approved):
Thanks for taking care of the randomness!

ArthurZucker committed:
improve context manager

Improve context manager's efficiency by filtering the forward calls base on the file origin

ArthurZucker commented:
I can confirm that it now fixes the tests for `vit_mae` as suggested by @NielsRogge. Just added a quick fix to only set the seed on the model's forward. LGTM

ArthurZucker commented on src/transformers/testing_utils.py:
```diff
@@ -1615,3 +1617,33 @@ def new_request(self, method, **kwargs):
+
+    def trace_calls(self, frame, event, arg):
+        # Set the seed when it is a call to a forward function from the model
+        if "/modeling_" in frame.f_code.co_filename:
```
BTW the frames looked a bit like that : 
```
<frame at 0x564b14f28360, file '/home/arthur_huggingface_co/transformers/src/transformers/models/vilt/modeling_vilt.py', line 761, code forward>
<code object forward at 0x7fe2c0610f50, file "/home/arthur_huggingface_co/transformers/src/transformers/models/vilt/modeling_vilt.py", line 761>
/home/arthur_huggingface_co/transformers/src/transformers/models/vilt/modeling_vilt.py
<frame at 0x564b14d71fe0, file '/home/arthur_huggingface_co/transformers/src/transformers/models/vilt/modeling_vilt.py', line 200, code forward>
<code object forward at 0x7fe2c060d0e0, file "/home/arthur_huggingface_co/transformers/src/transformers/models/vilt/modeling_vilt.py", line 200>
/home/arthur_huggingface_co/transformers/src/transformers/models/vilt/modeling_vilt.py
<frame at 0x564b1432e910, file '/home/arthur_huggingface_co/transformers/src/transformers/models/vilt/modeling_vilt.py', line 265, code forward>
<code object forward at 0x7fe2c060d2f0, file "/home/arthur_huggingface_co/transformers/src/transformers/models/vilt/modeling_vilt.py", line 265>
/home/arthur_huggingface_co/transformers/src/transformers/models/vilt/modeling_vilt.py
<frame at 0x564b14ce68b0, file '/opt/conda/envs/jukebox/lib/python3.8/site-packages/torch/nn/modules/sparse.py', line 157, code forward>
<code object forward at 0x7fe40cd5dc90, file "/opt/conda/envs/jukebox/lib/python3.8/site-packages/torch/nn/modules/sparse.py", line 157>
/opt/conda/envs/jukebox/lib/python3.8/site-packages/torch/nn/modules/sparse.py
<frame at 0x564b14ce68b0, file '/opt/conda/envs/jukebox/lib/python3.8/site-packages/torch/nn/modules/sparse.py', line 157, code forward>
<code object forward at 0x7fe40cd5dc90, file "/opt/conda/envs/jukebox/lib/python3.8/site-packages/torch/nn/modules/sparse.py", line 157>
/opt/conda/envs/jukebox/lib/python3.8/site-packages/torch/nn/modules/sparse.py
<frame at 0x564b14ce68b0, file '/opt/conda/envs/jukebox/lib/python3.8/site-packages/torch/nn/modules/sparse.py', line 157, code forward>
<code object forward at 0x7fe40cd5dc90, file "/opt/conda/envs/jukebox/lib/python3.8/site-packages/torch/nn/modules/sparse.py", line 157>
/opt/conda/envs/jukebox/lib/python3.8/site-packages/torch/nn/modules/sparse.py
```

younesbelkada commented on src/transformers/testing_utils.py:
```diff
@@ -1615,3 +1617,33 @@ def new_request(self, method, **kwargs):
+    def set_seed(self, frame, event, arg):
+        # Set the seed
+        set_seed(self.seed)
+        return
```
```suggestion
```

ArthurZucker committed:
Update src/transformers/testing_utils.py

small nit

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

stas00 commented on src/transformers/testing_utils.py:
```diff
@@ -1615,3 +1617,32 @@ def new_request(self, method, **kwargs):
+    def trace_calls(self, frame, event, arg):
+        # Set the seed when it is a call to a forward function from the model
+        if "/modeling_" in frame.f_code.co_filename:
+            return self.set_seed
```
So if I understand it correctly, you're using this hack to force the same seed before every python call, correct?

What is it that you're testing then, since this is not the normal behavior, i.e. normal code base sets the seed once to be reproducible.

In other words why setting the seed at the beginning of the program doesn't lead to a reproducible outcome? Isn't that an indicator of a bug in the proposed software and this hack tries to hide it?

and if you're really going to keep this can we call it something with a better mnemonic like `with reset_seed_on_every_frame`? Since `set_reproducible` to me means setting a seed once at the enter and leaving it alone.

younesbelkada replied:
Thanks @stas00 for your reply! 

Regarding your first point yes but, we set the seed before each `forward` call from a Pytorch module instead of any python call (from @ArthurZucker 's modification).

Regarding your second point, I was also very confused at the beginning but setting the pytorch seed before each `forward` seems to be needed to ensure reproducibility for stochastic operations. E.g. this code snippet shows that:
```
>>> import torch
>>> torch.manual_seed(0)
<torch._C.Generator object at 0x10576b6f0>
>>> torch.randn(1, 3)
tensor([[ 1.5410, -0.2934, -2.1788]])
>>> torch.randn(1, 3)
tensor([[ 0.5684, -1.0845, -1.3986]])
>>> torch.manual_seed(0)
<torch._C.Generator object at 0x10576b6f0>
>>> torch.randn(1, 3)
tensor([[ 1.5410, -0.2934, -2.1788]])
```
You need to set the seed before each stochastic call to ensure reproducibility. From my understanding that is why we need to call the seed setting on stochastic tests (e.g. [here](https://github.com/huggingface/transformers/blob/0f257a87749e0a72bda260c6f319a45dae1e7c4d/tests/models/vit_mae/test_modeling_vit_mae.py#L235) where the seed is called before each forward pass. I am still not sure why calling the seed once is not sufficient but I think that it might be because the seed has been 'consumed' when a stochastic operation is performed (This has been also observed on Jukebox integration I think).

For the last point yes! I think that we should change the decorator's name to make it more understandable

stas00 replied:
Ah, I missed that nuance. Then `reset_seed_before_every_forward` would be more appropriate.

But there is already a mechanism for doing this - this is called `register_forward_hook`
https://pytorch.org/docs/stable/generated/torch.jit.ScriptModule.html?highlight=hook#torch.jit.ScriptModule.register_forward_hook

so you don't need any cool hacks to accomplish that.

e.g. see:

https://github.com/huggingface/transformers/blob/4c2e983f44ce4d3b9c8502d42cc568e45897bd15/src/transformers/debug_utils.py#L160

ArthurZucker replied:
Very interesting, both option work, and maybe `register_forward_pre_hook(hook)` is more appropriate. 
IMO having a decorator will allow anyone to use it without modifying their test function, but we might have a lot of decorators. 

We could define the following
```python
# testing_utils.py
def pre_hook_set_seed(seed):
    def pre_hook(module, input):
        set_seed(seed)
    return pre_hook
```
And call 
```python 
# test_modeling_vilt.py
handle = self.register_forward_pre_hook(pre_hook_set_seed)
...
handle.remove()
```
I guess its pretty similar but requires a bit more code. 😄 
@stas00 What do you think would work best long term?

stas00 replied:
the pytorch call is designated for this specific functionality so it's a much cleaner solution, IMHO.

You can just as well implement it as a container so enter/exit for the above sample.

Also I was thinking that if ViLT needs this functionality often, perhaps the hook should be added directly into its modeling file?

But testing_utils is perfect too if it's generic and other models might need it too.

ArthurZucker replied:
Agreed! I will try to make a decorator using the `ur f`. 
Thanks for your feedbacks

ArthurZucker replied:
Hey! So, the problem with the pytorch hook is that you need to grab the module instance in order to link the hook. Thus you can't really just have a decorator for the test, you need to re-write the test (as the models are accessed by their names), or maybe I was just unable to do so 😅 . 

Here is the final solution : 

```python 
    def trace_calls(self, frame, event, arg):
        # frame.f_locals provides the arguments of the function that was called 
        # we just check that it is a ...Model class 
        with suppress(KeyError):
            if "Model" in str(frame.f_locals["self"].__class__):
                return self.set_seed
``` 
@sgugger  What do you think?

sgugger replied:
This seems a bit overengineered for three tests where we could just insert two `set_seed` calls to ensure the same results (even registering a pre-foward hook seems over-engineered TBH). Or am I missing something?

ArthurZucker replied:
You are right 😅 But since `VitMae` also requires that hack, we though that long term, made more sense to have an easy to use decorator! You have to insert the `set_seed` before every `forward` (so in the for loop that is testing) for each of the test so the original idea to have a decorator followed Yih-dar's comment. 
Anyway will do as you think works better!

sgugger replied:
While I was excited by the idea at first, reading Stas' comments and thinking more about it, I think we should stick to something simple :-)

stas00 replied:
Why does it have to be a test decorator? what would be wrong with calling it from the test?

Also I was thinking that if this might be used by users - e.g. to debug their code, won't it be a good idea to make this functionality as part of the model itself?

So basically one would call:

```
model = from_pretrained(...)
model.reset_seed_before_every_forward(True)
``` 

and so now it's accessible to tests and users too.

If multiple models require that it could be a mixin class and they can inherit from it, so it won't make any noise in the modeling file.

sgugger replied:
Also fine with this.

ArthurZucker committed:
update

ArthurZucker committed:
style

ArthurZucker committed:
supress KeyError

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into vilt-accelerate-suuport

younesbelkada committed:
remove unused import

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into vilt-accelerate-suuport

younesbelkada committed:
major changes

- remove `reset_seed_before_every_forward` context manager
- redefine accelerate tests

younesbelkada committed:
revert local image

younesbelkada commented:
Thank you all for your comments!
Reverted the changes related to the context manager and kept the tests to be as simple as possible!
Can confirm all the slow tests pass now (expect for [this test](https://github.com/huggingface/transformers/pull/18683#discussion_r949422087) but this is expected as stated in the comment)
I propose to address the stochasticity issue potentially in a follow-up PR

sgugger reviewed (commented):
It looks like the overloaded tests are just adding two lines for setting the seed. That modification can definitely goo in the base test (sorry if I was unclear in my comments before) :-)

younesbelkada committed:
add new attribute on `accelerate` tests

- add the attribute `reset_seed_before_every_forward` on `accelerate` tests
- this supports stochastic tests

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into vilt-accelerate-suuport

younesbelkada committed:
add `cached_property`

younesbelkada commented:
@sgugger thanks for your comment! And sorry for my late reply
I should have addressed the proposed suggestion in the commit 43b087a ! I hesitated between this solution and having an attribute on `ModelTesterMixin` but I thought that this solution is simpler to understand for future users (the `ModelTesterMixin` has already a lot of attributes)

sgugger reviewed (commented):
I didn't mean adding a new argument to the test. We can have the seed set in the tests for all models. :-)

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
some final changes

- set seed before every forward on `accelerate` tests
- `make fixup`

younesbelkada commented:
Perfect thanks! @sgugger 
Should have addressed the changes now! Let me know if you still need some modifications

sgugger commented on tests/models/vilt/test_modeling_vilt.py:
```diff
@@ -509,10 +516,24 @@ def test_model_from_pretrained(self):
             model = ViltModel.from_pretrained(model_name)
             self.assertIsNotNone(model)
 
+    @require_accelerate
+    @require_torch_gpu
+    def test_cpu_offload(self):
+        super().test_cpu_offload()
+
+    @require_accelerate
+    @require_torch_gpu
+    def test_disk_offload(self):
+        super().test_disk_offload()
+
+    @require_accelerate
+    @require_torch_multi_gpu
+    def test_model_parallelism(self):
+        super().test_model_parallelism()
```
None of this is needed anymore ;-p

younesbelkada replied:
Oops yes ahah forgot to remove them

younesbelkada committed:
remove `accelerate` tests

- tests will be correctly called from the base class

sgugger reviewed (approved):
Thanks a lot!
